app/src/org/commcare/dalvik/activities/CommCareHomeActivity.java

```diff
@@ -270,149 +387,166 @@ protected void onActivityResult(int requestCode, int resultCode, Intent intent)
+                        formEntry(platform.getFormContentUri(r.getFormNamespace()), r);
+                        return;
+                    }
+                    break;
+                case GET_COMMAND:
+                    //TODO: We might need to load this from serialized state?
+                    AndroidSessionWrapper currentState =
+                            CommCareApplication._().getCurrentSessionWrapper();
+                    if (resultCode == RESULT_CANCELED) {
+                        if (currentState.getSession().getCommand() == null) {
+                            //Needed a command, and didn't already have one. Stepping back from
+                            //an empty state, Go home!
+                            currentState.reset();
+                            uiController.refreshView();
+                            return;
+                        } else {
+                            currentState.getSession().stepBack();
+                        }
+                    } else if (resultCode == RESULT_OK) {
+                        CommCareSession session = currentState.getSession();
+                        if (sessionStateUnchangedSinceCallout(session, intent)) {
+                            //Get our command, set it, and continue forward
+                            String command = intent.getStringExtra(SessionFrame.STATE_COMMAND_ID);
+                            session.setCommand(command);
+                        } else {
+                            clearSessionAndExit(currentState, true);
+                            return;
+                        }
+                    }
+                    break;
+                case GET_CASE:
+                    //TODO: We might need to load this from serialized state?
+                    AndroidSessionWrapper asw = CommCareApplication._().getCurrentSessionWrapper();
+                    CommCareSession currentSession = asw.getSession();
+                    if (resultCode == RESULT_CANCELED) {
+                        currentSession.stepBack();
+                    } else if (resultCode == RESULT_OK) {
+                        if (sessionStateUnchangedSinceCallout(currentSession, intent)) {
+                            String sessionDatumId = currentSession.getNeededDatum().getDataId();
+                            String chosenCaseId = intent.getStringExtra(SessionFrame.STATE_DATUM_VAL);
+                            currentSession.setDatum(sessionDatumId, chosenCaseId);
+                        } else {
+                            clearSessionAndExit(asw, true);
+                            return;
+                        }
+                    }
+                    break;
+                case MODEL_RESULT:
+                    boolean fetchNext = processReturnFromFormEntry(resultCode, intent);
+                    if (!fetchNext) {
+                        return;
+                    }
+                    break;
+                case AUTHENTICATION_FOR_PIN:
```

Comment:
@ctsims Yea I was concerned about that initially but talked with Phillip about it and neither of us could really come up with a better strategy. The passing of the un-hashed password here and in the other place you flagged is obviously essential to being able to set a PIN. What alternative would you suggest?